File: 3.无重复字符的最长子串.py
#
# @lc app=leetcode.cn id=3 lang=python3
#
# [3] 无重复字符的最长子串
#

# @lc code=start
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if len(s) == 0:
            return 0
        length = 0
        record = {}
        i, j, n, max_length = 0, 1, 1, 1
        record[s[i]] = i
        while j < len(s):
            if s[j] in record:
                while s[j] in record:
                    del record[s[i]]
                    i += 1
                    n -= 1
            
            record[s[j]] = j
            n += 1
            j += 1
            if n > max_length:
                max_length = n
        return max_length
# @lc code=end

File: 3.无重复字符的最长子串.cpp
/*
 * @lc app=leetcode.cn id=3 lang=cpp
 *
 * [3] 无重复字符的最长子串
 */

// @lc code=start
#include<map>
using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int min = 0;
        map<char, int> stored;
        int i = 0, j = 0;
        // valid segment [i,j)
        while(j < s.size()){
            if(stored.find(s[j]) == stored.end()){
                // next char ok to be added
                stored[s[j]] = j;
                j++;
            }else{
                stored.erase(stored.find(s[i]));
                i++;
            }
            if(min < j - i) min = j - i;
        }
        return min;
    }
};
// @lc code=end

